File: ports/zig/src/layout.zig
//! Sizes are resolved once per frame with a single pass. No manual coordinate
//! arithmetic should ever appear in application code.
//!
//! ```
//! Size.cells(12)     12 columns/rows
//! Size.percent(40)   40% of the container
//! Size.fr(2)         two shares of whatever is left over
//! Size.auto          whatever the widget says it needs
//! Size.fill          same as fr(1)
//! ```
//!
//! The reference spells these as numbers and strings ("2fr", "40%"). Both work
//! here: `Size.cells(12)`, or `Size.parse("2fr")`.

const std = @import("std");

const roundHalfUp = @import("color.zig").roundHalfUp;

/// How much space an item wants along the main axis.
pub const Size = union(enum) {
    cells: i64,
    percent: f64,
    fr: f64,
    auto,
    fill,

    /// Parse the string spelling the reference API uses. Anything
    /// unrecognisable becomes zero cells, matching `Number.parseFloat`'s NaN
    /// fallback there.
    pub fn parse(spec: []const u8) Size {
        const s = std.mem.trim(u8, spec, " \t\r\n");
        if (std.mem.eql(u8, s, "auto")) return .auto;
        if (std.mem.eql(u8, s, "fill")) return .fill;
        if (s.len > 0 and s[s.len - 1] == '%') {
            return .{ .percent = leadingFloat(s[0 .. s.len - 1]) };
        }
        if (s.len > 2 and std.mem.endsWith(u8, s, "fr")) {
            const n = leadingFloat(s[0 .. s.len - 2]);
            return .{ .fr = if (n > 0) n else 1 };
        }
        return .{ .cells = @intFromFloat(leadingFloat(s)) };
    }
};

/// `Number.parseFloat`: read as much of a leading number as parses.
fn leadingFloat(text: []const u8) f64 {
    const s = std.mem.trim(u8, text, " \t\r\n");
    var end: usize = 0;
    var seen_digit = false;
    var seen_dot = false;
    while (end < s.len) : (end += 1) {
        const c = s[end];
        if ((c == '+' or c == '-') and end == 0) continue;
        if (c >= '0' and c <= '9') {
            seen_digit = true;
            continue;
        }
        if (c == '.' and !seen_dot) {
            seen_dot = true;
            continue;
        }
        break;
    }
    if (!seen_digit) return 0;
    return std.fmt.parseFloat(f64, s[0..end]) catch 0;
}

/// What one item contributes to a layout solve.
pub const Constraint = struct {
    size: ?Size = null,
    min: ?usize = null,
    max: ?usize = null,
    /// Natural size, used by `auto` and as the floor for flexible items.
    intrinsic: ?usize = null,

    pub fn sized(s: Size) Constraint {
        return .{ .size = s };
    }

    /// The reference's `minmax(min, max)` helper.
    pub fn minMax(lo: usize, hi: usize) Constraint {
        return .{ .size = .fill, .min = lo, .max = hi };
    }
};

/// A rectangle in absolute buffer coordinates.
pub const Rect = struct {
    x: isize = 0,
    y: isize = 0,
    width: usize = 0,
    height: usize = 0,

    pub fn isEmpty(self: Rect) bool {
        return self.width == 0 or self.height == 0;
    }

    pub fn contains(self: Rect, x: isize, y: isize) bool {
        return x >= self.x and y >= self.y and
            x < self.x + @as(isize, @intCast(self.width)) and
            y < self.y + @as(isize, @intCast(self.height));
    }

    /// Shrink by padding, never past zero.
    pub fn inset(self: Rect, padding: Padding) Rect {
        return .{
            .x = self.x + @as(isize, @intCast(padding.left)),
            .y = self.y + @as(isize, @intCast(padding.top)),
            .width = self.width -| padding.left -| padding.right,
            .height = self.height -| padding.top -| padding.bottom,
        };
    }

    pub fn intersect(self: Rect, other: Rect) Rect {
        const x = @max(self.x, other.x);
        const y = @max(self.y, other.y);
        const x2 = @min(
            self.x + @as(isize, @intCast(self.width)),
            other.x + @as(isize, @intCast(other.width)),
        );
        const y2 = @min(
            self.y + @as(isize, @intCast(self.height)),
            other.y + @as(isize, @intCast(other.height)),
        );
        return .{
            .x = x,
            .y = y,
            .width = @intCast(@max(0, x2 - x)),
            .height = @intCast(@max(0, y2 - y)),
        };
    }
};

/// Padding, clockwise from the top.
pub const Padding = struct {
    top: usize = 0,
    right: usize = 0,
    bottom: usize = 0,
    left: usize = 0,

    pub const none: Padding = .{};

    pub fn all(v: usize) Padding {
        return .{ .top = v, .right = v, .bottom = v, .left = v };
    }

    pub fn axes(vertical: usize, horizontal: usize) Padding {
        return .{ .top = vertical, .right = horizontal, .bottom = vertical, .left = horizontal };
    }
};

/// Which way a container lays its children out.
pub const Direction = enum { row, column };

const Resolved = struct {
    value: f64,
    fr: f64,
    min: f64,
    max: f64,
};

fn parseConstraint(c: Constraint, total: usize) Resolved {
    const lo: f64 = if (c.min) |v| @floatFromInt(v) else 0;
    const hi: f64 = if (c.max) |v| @floatFromInt(v) else std.math.inf(f64);
    const size = c.size orelse .auto;
    return switch (size) {
        .cells => |n| .{ .value = @floatFromInt(n), .fr = 0, .min = lo, .max = hi },
        .auto => .{
            .value = if (c.intrinsic) |v| @floatFromInt(v) else 0,
            .fr = 0,
            .min = lo,
            .max = hi,
        },
        .fill => .{ .value = 0, .fr = 1, .min = lo, .max = hi },
        .percent => |p| blk: {
            var pct = p / 100;
            if (!std.math.isFinite(pct)) pct = 0;
            break :blk .{
                .value = roundHalfUp(@as(f64, @floatFromInt(total)) * pct),
                .fr = 0,
                .min = lo,
                .max = hi,
            };
        },
        .fr => |n| .{
            .value = 0,
            .fr = if (std.math.isFinite(n) and n > 0) n else 1,
            .min = lo,
            .max = hi,
        },
    };
}

fn clampF(v: f64, lo: f64, hi: f64) f64 {
    return @max(lo, @min(hi, v));
}

/// Distribute `total` across `items`, honouring gaps, fractions and min/max.
/// Always returns non-negative sizes that sum to at most `total`. The caller
/// owns the returned slice.
pub fn solve(
    allocator: std.mem.Allocator,
    total: usize,
    items: []const Constraint,
    gap: usize,
) ![]usize {
    const seams = try allocator.alloc(isize, items.len -| 1);
    defer allocator.free(seams);
    @memset(seams, @intCast(gap));
    return solveWithGaps(allocator, total, items, seams);
}

/// As `solve`, but with a gap per seam, which may be negative.
///
/// A negative seam is how collapsed borders work: two panels overlap by the
/// column their borders share, so the pair occupies one column less than the
/// sum of their widths.
pub fn solveWithGaps(
    allocator: std.mem.Allocator,
    total: usize,
    items: []const Constraint,
    gaps: []const isize,
) ![]usize {
    const n = items.len;
    if (n == 0) return allocator.alloc(usize, 0);

    var gap_total: isize = 0;
    for (gaps[0..@min(gaps.len, n - 1)]) |g| gap_total += g;
    const signed_available = @as(isize, @intCast(total)) - gap_total;
    const available: usize = if (signed_available > 0) @intCast(signed_available) else 0;
    const parsed = try allocator.alloc(Resolved, n);
    defer allocator.free(parsed);
    for (items, 0..) |c, i| parsed[i] = parseConstraint(c, available);

    // -1 marks an item resolved in the flexible pass below.
    const out = try allocator.alloc(i64, n);
    defer allocator.free(out);
    @memset(out, -1);

    var used: i64 = 0;
    var fr_total: f64 = 0;
    for (0..n) |i| {
        const p = parsed[i];
        if (p.fr > 0) {
            fr_total += p.fr;
        } else {
            const v = clampF(
                roundHalfUp(p.value),
                p.min,
                @min(p.max, @as(f64, @floatFromInt(available))),
            );
            out[i] = @intFromFloat(v);
            used += out[i];
        }
    }

    const free = @max(0.0, @as(f64, @floatFromInt(@as(i64, @intCast(available)) - used)));
    if (fr_total > 0) {
        // Two passes: clamped items give their surplus back to the rest.
        var remaining_fr = fr_total;
        var pool = free;
        var pending = try std.ArrayList(usize).initCapacity(allocator, n);
        defer pending.deinit(allocator);
        for (0..n) |i| {
            if (out[i] == -1) pending.appendAssumeCapacity(i);
        }

        var changed = true;
        while (changed and pending.items.len > 0) {
            changed = false;
            var idx: usize = 0;
            while (idx < pending.items.len) {
                const i = pending.items[idx];
                const p = parsed[i];
                const share = if (remaining_fr > 0) pool * p.fr / remaining_fr else 0;
                const clamped = clampF(share, p.min, p.max);
                if (clamped != share) {
                    out[i] = @intFromFloat(roundHalfUp(clamped));
                    pool -= @floatFromInt(out[i]);
                    remaining_fr -= p.fr;
                    _ = pending.orderedRemove(idx);
                    changed = true;
                } else {
                    idx += 1;
                }
            }
        }

        // Distribute what is left, giving the rounding remainder to the last.
        var assigned: i64 = 0;
        for (pending.items, 0..) |i, k| {
            const p = parsed[i];
            const exact = if (remaining_fr > 0) pool * p.fr / remaining_fr else 0;
            const v: i64 = if (k == pending.items.len - 1)
                @max(0, @as(i64, @intFromFloat(pool)) - assigned)
            else
                @intFromFloat(@floor(exact));
            out[i] = v;
            assigned += v;
        }
    }

    // Overflow: shrink from the end until it fits rather than drawing outside.
    var sum: i64 = 0;
    for (out) |v| sum += v;
    const avail: i64 = @intCast(available);
    if (sum > avail) {
        var i: usize = n;
        while (i > 0 and sum > avail) {
            i -= 1;
            const shrink = @min(out[i] - @as(i64, @intFromFloat(parsed[i].min)), sum - avail);
            if (shrink > 0) {
                out[i] -= shrink;
                sum -= shrink;
            }
        }
        i = n;
        while (i > 0 and sum > avail) {
            i -= 1;
            const shrink = @min(out[i], sum - avail);
            out[i] -= shrink;
            sum -= shrink;
        }
    }

    const result = try allocator.alloc(usize, n);
    for (out, 0..) |v, i| result[i] = @intCast(@max(0, v));
    return result;
}

/// Lay children out along one axis inside `rect`. The caller owns the slice.
/// Where leftover space goes.
///
/// It only ever applies when there is slack, and a container holding any `fr`
/// or `fill` child has none -- that child has already absorbed it. So this is
/// inert exactly where it would otherwise fight with the constraints.
pub const Justify = enum {
    start,
    end,
    center,
    space_between,
    space_around,
    space_evenly,

    /// The spelling the reference API uses. Anything else is `start`, which is
    /// what every layout did before this existed.
    pub fn parse(name: []const u8) Justify {
        if (std.mem.eql(u8, name, "end")) return .end;
        if (std.mem.eql(u8, name, "center")) return .center;
        if (std.mem.eql(u8, name, "space-between")) return .space_between;
        if (std.mem.eql(u8, name, "space-around")) return .space_around;
        if (std.mem.eql(u8, name, "space-evenly")) return .space_evenly;
        return .start;
    }
};

/// How much slack sits before item `i`, as an exact fraction. Every mode is a
/// different answer to that one question, which is why they share the rounding
/// below rather than each growing their own off-by-one.
fn before(i: usize, slack: f64, count: usize, justify: Justify) f64 {
    const fi: f64 = @floatFromInt(i);
    const fc: f64 = @floatFromInt(count);
    return switch (justify) {
        .start => 0,
        .end => slack,
        // Floor, so an odd cell falls after the content rather than before it.
        .center => @floor(slack / 2),
        .space_between => if (count > 1) fi * slack / (fc - 1) else 0,
        .space_evenly => (fi + 1) * slack / (fc + 1),
        .space_around => (fi + 0.5) * slack / fc,
    };
}

/// The offset before the first child, and the extra added at each seam. The
/// caller owns the returned slice.
///
/// Cells are whole, and rounding each gap on its own loses one here and gains
/// one there. Rounding the cumulative offset and taking differences means the
/// parts always add up to exactly the slack.
pub fn distribute(
    allocator: std.mem.Allocator,
    slack: usize,
    count: usize,
    justify: Justify,
) !struct { lead: usize, seams: []usize } {
    const seams = try allocator.alloc(usize, count -| 1);
    @memset(seams, 0);
    if (slack == 0 or count == 0 or justify == .start) return .{ .lead = 0, .seams = seams };

    const at = struct {
        fn f(i: usize, sl: usize, c: usize, j: Justify) usize {
            const v = before(i, @floatFromInt(sl), c, j);
            return @intFromFloat(@floor(v + 0.5));
        }
    }.f;

    var i: usize = 0;
    while (i + 1 < count) : (i += 1) {
        seams[i] = at(i + 1, slack, count, justify) -| at(i, slack, count, justify);
    }
    return .{ .lead = at(0, slack, count, justify), .seams = seams };
}

pub fn stack(
    allocator: std.mem.Allocator,
    rect: Rect,
    items: []const Constraint,
    direction: Direction,
    gap: usize,
) ![]Rect {
    const seams = try allocator.alloc(isize, items.len -| 1);
    defer allocator.free(seams);
    @memset(seams, @intCast(gap));
    return stackWithGaps(allocator, rect, items, direction, seams);
}

/// As `stack`, with a policy for whatever the children leave over.
pub fn stackJustified(
    allocator: std.mem.Allocator,
    rect: Rect,
    items: []const Constraint,
    direction: Direction,
    gap: usize,
    justify: Justify,
) ![]Rect {
    const seams = try allocator.alloc(isize, items.len -| 1);
    defer allocator.free(seams);
    @memset(seams, @intCast(gap));
    return stackWithGapsJustified(allocator, rect, items, direction, seams, justify);
}

/// As `stack`, but with a gap per seam, which may be negative.
pub fn stackWithGaps(
    allocator: std.mem.Allocator,
    rect: Rect,
    items: []const Constraint,
    direction: Direction,
    gaps: []const isize,
) ![]Rect {
    return stackWithGapsJustified(allocator, rect, items, direction, gaps, .start);
}

/// The full form: a gap per seam, and a policy for whatever is left over.
pub fn stackWithGapsJustified(
    allocator: std.mem.Allocator,
    rect: Rect,
    items: []const Constraint,
    direction: Direction,
    gaps: []const isize,
    justify: Justify,
) ![]Rect {
    const horizontal = direction == .row;
    const axis = if (horizontal) rect.width else rect.height;
    const sizes = try solveWithGaps(allocator, axis, items, gaps);
    defer allocator.free(sizes);

    var gap_total: isize = 0;
    var g: usize = 0;
    while (g + 1 < sizes.len) : (g += 1) {
        if (g < gaps.len) gap_total += gaps[g];
    }
    var used: isize = gap_total;
    for (sizes) |size| used += @intCast(size);
    const slack: usize = @intCast(@max(0, @as(isize, @intCast(axis)) - used));

    const spread = try distribute(allocator, slack, sizes.len, justify);
    defer allocator.free(spread.seams);

    const out = try allocator.alloc(Rect, sizes.len);
    var offset: isize = (if (horizontal) rect.x else rect.y) + @as(isize, @intCast(spread.lead));
    for (sizes, 0..) |size, i| {
        out[i] = if (horizontal)
            .{ .x = offset, .y = rect.y, .width = size, .height = rect.height }
        else
            .{ .x = rect.x, .y = offset, .width = rect.width, .height = size };
        const seam = (if (i < gaps.len) gaps[i] else 0) +
            @as(isize, @intCast(if (i < spread.seams.len) spread.seams[i] else 0));
        offset += @as(isize, @intCast(size)) + seam;
    }
    return out;
}

File: ports/rust/src/layout.rs
//! Sizes are resolved once per frame with a single pass. No manual coordinate
//! arithmetic should ever appear in application code.
//!
//! ```text
//! Size::Cells(12)     12 columns/rows
//! Size::Percent(40.0) 40% of the container
//! Size::Fr(2.0)       two shares of whatever is left over
//! Size::Auto          whatever the widget says it needs
//! Size::Fill          same as Fr(1.0)
//! ```
//!
//! The reference implementation spells these as numbers and strings (`"2fr"`,
//! `"40%"`). Both spellings work here: `Size::from(12)`, `Size::from("2fr")`.

use crate::color::round_half_up;

/// How much space an item wants along the main axis.
#[derive(Clone, Copy, Debug, PartialEq)]
pub enum Size {
    Cells(i64),
    Percent(f64),
    Fr(f64),
    Auto,
    Fill,
}

impl Default for Size {
    fn default() -> Size {
        Size::Auto
    }
}

impl Size {
    /// Parse the string spelling the TypeScript API uses. Anything
    /// unrecognisable becomes zero cells, matching `Number.parseFloat`'s NaN
    /// fallback there.
    pub fn parse(spec: &str) -> Size {
        let s = spec.trim();
        if s == "auto" {
            return Size::Auto;
        }
        if s == "fill" {
            return Size::Fill;
        }
        if let Some(rest) = s.strip_suffix('%') {
            return Size::Percent(rest.trim().parse::<f64>().unwrap_or(0.0));
        }
        if let Some(rest) = s.strip_suffix("fr") {
            let n = rest.trim().parse::<f64>().unwrap_or(1.0);
            return Size::Fr(if n > 0.0 { n } else { 1.0 });
        }
        Size::Cells(leading_float(s).unwrap_or(0.0) as i64)
    }
}

/// `Number.parseFloat` semantics: read as much of a leading number as parses.
fn leading_float(s: &str) -> Option<f64> {
    let mut end = 0usize;
    let bytes = s.as_bytes();
    let mut seen_digit = false;
    let mut seen_dot = false;
    while end < bytes.len() {
        let c = bytes[end] as char;
        match c {
            '+' | '-' if end == 0 => {}
            '0'..='9' => seen_digit = true,
            '.' if !seen_dot => seen_dot = true,
            _ => break,
        }
        end += 1;
    }
    if !seen_digit {
        return None;
    }
    s[..end].parse::<f64>().ok()
}

impl From<i64> for Size {
    fn from(n: i64) -> Size {
        Size::Cells(n)
    }
}

impl From<i32> for Size {
    fn from(n: i32) -> Size {
        Size::Cells(n as i64)
    }
}

impl From<usize> for Size {
    fn from(n: usize) -> Size {
        Size::Cells(n as i64)
    }
}

impl From<&str> for Size {
    fn from(s: &str) -> Size {
        Size::parse(s)
    }
}

/// What one item contributes to a layout solve.
#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub struct Constraint {
    pub size: Option<Size>,
    pub min: Option<usize>,
    pub max: Option<usize>,
    /// Natural size, used by `Auto` and as the floor for flexible items.
    pub intrinsic: Option<usize>,
}

impl Constraint {
    pub fn new(size: impl Into<Size>) -> Constraint {
        Constraint { size: Some(size.into()), ..Default::default() }
    }

    pub fn fixed(n: usize) -> Constraint {
        Constraint::new(Size::Cells(n as i64))
    }

    pub fn fill() -> Constraint {
        Constraint::new(Size::Fill)
    }

    pub fn flex(n: f64) -> Constraint {
        Constraint::new(Size::Fr(n))
    }

    pub fn auto(intrinsic: usize) -> Constraint {
        Constraint { size: Some(Size::Auto), intrinsic: Some(intrinsic), ..Default::default() }
    }

    pub fn minmax(min: usize, max: usize) -> Constraint {
        Constraint {
            size: Some(Size::Fill),
            min: Some(min),
            max: Some(max),
            intrinsic: None,
        }
    }

    pub fn with_min(mut self, min: usize) -> Constraint {
        self.min = Some(min);
        self
    }

    pub fn with_max(mut self, max: usize) -> Constraint {
        self.max = Some(max);
        self
    }

    pub fn with_intrinsic(mut self, intrinsic: usize) -> Constraint {
        self.intrinsic = Some(intrinsic);
        self
    }
}

/// A rectangle in absolute buffer coordinates.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub struct Rect {
    pub x: isize,
    pub y: isize,
    pub width: usize,
    pub height: usize,
}

impl Rect {
    pub const fn new(x: isize, y: isize, width: usize, height: usize) -> Rect {
        Rect { x, y, width, height }
    }

    pub const fn is_empty(&self) -> bool {
        self.width == 0 || self.height == 0
    }

    pub fn contains(&self, x: isize, y: isize) -> bool {
        x >= self.x
            && y >= self.y
            && x < self.x + self.width as isize
            && y < self.y + self.height as isize
    }

    /// Shrink by padding, never past zero.
    pub fn inset(&self, padding: Padding) -> Rect {
        let (t, r, b, l) = padding.resolve();
        Rect {
            x: self.x + l as isize,
            y: self.y + t as isize,
            width: self.width.saturating_sub(l + r),
            height: self.height.saturating_sub(t + b),
        }
    }

    pub fn intersect(&self, other: Rect) -> Rect {
        let x = self.x.max(other.x);
        let y = self.y.max(other.y);
        let x2 = (self.x + self.width as isize).min(other.x + other.width as isize);
        let y2 = (self.y + self.height as isize).min(other.y + other.height as isize);
        Rect {
            x,
            y,
            width: (x2 - x).max(0) as usize,
            height: (y2 - y).max(0) as usize,
        }
    }
}

/// Padding in the CSS shorthand orders: one value, vertical/horizontal, or all
/// four clockwise from the top.
#[derive(Clone, Copy, Debug, PartialEq, Eq, Default)]
pub enum Padding {
    #[default]
    None,
    All(usize),
    /// `(vertical, horizontal)`
    Axes(usize, usize),
    /// `(top, right, bottom, left)`
    Sides(usize, usize, usize, usize),
}

impl Padding {
    pub const fn resolve(self) -> (usize, usize, usize, usize) {
        match self {
            Padding::None => (0, 0, 0, 0),
            Padding::All(v) => (v, v, v, v),
            Padding::Axes(v, h) => (v, h, v, h),
            Padding::Sides(t, r, b, l) => (t, r, b, l),
        }
    }
}

impl From<usize> for Padding {
    fn from(v: usize) -> Padding {
        Padding::All(v)
    }
}

impl From<(usize, usize)> for Padding {
    fn from(v: (usize, usize)) -> Padding {
        Padding::Axes(v.0, v.1)
    }
}

/// Which way a container lays its children out.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Direction {
    Row,
    Column,
}

#[derive(Clone, Copy)]
struct Resolved {
    value: f64,
    fr: f64,
    min: f64,
    max: f64,
}

fn parse_constraint(c: &Constraint, total: usize) -> Resolved {
    let min = c.min.unwrap_or(0) as f64;
    let max = c.max.map(|m| m as f64).unwrap_or(f64::INFINITY);
    match c.size.unwrap_or(Size::Auto) {
        Size::Cells(n) => Resolved { value: n as f64, fr: 0.0, min, max },
        Size::Auto => Resolved { value: c.intrinsic.unwrap_or(0) as f64, fr: 0.0, min, max },
        Size::Fill => Resolved { value: 0.0, fr: 1.0, min, max },
        Size::Percent(p) => {
            let pct = p / 100.0;
            let pct = if pct.is_finite() { pct } else { 0.0 };
            Resolved { value: round_half_up(total as f64 * pct), fr: 0.0, min, max }
        }
        Size::Fr(n) => Resolved {
            value: 0.0,
            fr: if n.is_finite() && n > 0.0 { n } else { 1.0 },
            min,
            max,
        },
    }
}

fn clamp(v: f64, min: f64, max: f64) -> f64 {
    v.min(max).max(min)
}

/// Distribute `total` across `items`, honouring gaps, fractions and min/max.
/// Always returns non-negative sizes that sum to at most `total`.
pub fn solve(total: usize, items: &[Constraint], gap: usize) -> Vec<usize> {
    let seams = vec![gap as isize; items.len().saturating_sub(1)];
    solve_with_gaps(total, items, &seams)
}

/// As `solve`, but with a gap per seam, which may be negative.
///
/// A negative seam is how collapsed borders work: two panels overlap by the
/// column their borders share, so the pair occupies one column less than the
/// sum of their widths.
pub fn solve_with_gaps(total: usize, items: &[Constraint], gaps: &[isize]) -> Vec<usize> {
    let n = items.len();
    if n == 0 {
        return Vec::new();
    }
    let gap_total: isize = gaps.iter().take(n - 1).sum();
    let available = (total as isize - gap_total).max(0) as usize;
    let parsed: Vec<Resolved> = items.iter().map(|c| parse_constraint(c, available)).collect();

    let mut used = 0i64;
    let mut fr_total = 0.0f64;
    // -1 marks an item resolved in the flexible pass below.
    let mut out = vec![-1i64; n];

    for i in 0..n {
        let p = parsed[i];
        if p.fr > 0.0 {
            fr_total += p.fr;
        } else {
            let v = clamp(round_half_up(p.value), p.min, p.max.min(available as f64));
            out[i] = v as i64;
            used += out[i];
        }
    }

    let free = (available as i64 - used).max(0) as f64;
    if fr_total > 0.0 {
        // Two passes: clamped items give their surplus back to the rest.
        let mut remaining_fr = fr_total;
        let mut pool = free;
        let mut pending: Vec<usize> = (0..n).filter(|&i| out[i] == -1).collect();

        let mut changed = true;
        while changed && !pending.is_empty() {
            changed = false;
            for i in pending.clone() {
                let p = parsed[i];
                let share = if remaining_fr > 0.0 { pool * p.fr / remaining_fr } else { 0.0 };
                let clamped = clamp(share, p.min, p.max);
                if clamped != share {
                    out[i] = round_half_up(clamped) as i64;
                    pool -= out[i] as f64;
                    remaining_fr -= p.fr;
                    pending.retain(|&j| j != i);
                    changed = true;
                }
            }
        }

        // Distribute what is left, giving the rounding remainder to the last.
        let mut assigned = 0i64;
        let count = pending.len();
        for (k, &i) in pending.iter().enumerate() {
            let p = parsed[i];
            let exact = if remaining_fr > 0.0 { pool * p.fr / remaining_fr } else { 0.0 };
            let v = if k == count - 1 {
                (pool as i64 - assigned).max(0)
            } else {
                exact.floor() as i64
            };
            out[i] = v;
            assigned += v;
        }
    }

    // Overflow: shrink from the end until it fits rather than drawing outside.
    let mut sum: i64 = out.iter().sum();
    let avail = available as i64;
    if sum > avail {
        for i in (0..n).rev() {
            if sum <= avail {
                break;
            }
            let shrink = (out[i] - parsed[i].min as i64).min(sum - avail);
            if shrink > 0 {
                out[i] -= shrink;
                sum -= shrink;
            }
        }
        for i in (0..n).rev() {
            if sum <= avail {
                break;
            }
            let shrink = out[i].min(sum - avail);
            out[i] -= shrink;
            sum -= shrink;
        }
    }

    out.into_iter().map(|v| v.max(0) as usize).collect()
}

/// Lay children out along one axis inside `rect`.
pub fn stack(rect: Rect, items: &[Constraint], direction: Direction, gap: usize) -> Vec<Rect> {
    let seams = vec![gap as isize; items.len().saturating_sub(1)];
    stack_with_gaps(rect, items, direction, &seams)
}

/// As `stack`, but with a gap per seam, which may be negative.
pub fn stack_with_gaps(
    rect: Rect,
    items: &[Constraint],
    direction: Direction,
    gaps: &[isize],
) -> Vec<Rect> {
    let horizontal = direction == Direction::Row;
    let sizes = solve_with_gaps(if horizontal { rect.width } else { rect.height }, items, gaps);
    let mut out = Vec::with_capacity(sizes.len());
    let mut offset = if horizontal { rect.x } else { rect.y };
    for (i, size) in sizes.into_iter().enumerate() {
        out.push(if horizontal {
            Rect { x: offset, y: rect.y, width: size, height: rect.height }
        } else {
            Rect { x: rect.x, y: offset, width: rect.width, height: size }
        });
        offset += size as isize + gaps.get(i).copied().unwrap_or(0);
    }
    out
}
